common/src/main/java/org/conscrypt/NativeSsl.java

```diff
@@ -655,27 +655,45 @@ private BioWrapper() throws SSLException {
         }
 
         int getPendingWrittenBytes() {
-            if (bio != 0) {
+            if (bio != 0L) {
                 return NativeCrypto.SSL_pending_written_bytes_in_BIO(bio);
             } else {
                 return 0;
             }
         }
 
         int writeDirectByteBuffer(long address, int length) throws IOException {
-            return NativeCrypto.ENGINE_SSL_write_BIO_direct(
-                    ssl, NativeSsl.this, bio, address, length, handshakeCallbacks);
+            lock.readLock().lock();
+            try {
+                if (isClosed()) {
+                    throw new SocketException("Socket is closed");
```

Comment:
1) That error message won't make sense for apps using `SSLEngine` directly.
2) I think these check (here and in `readDirect...`) should be at the `ConscryptEngine` level, both methods are only called from a single place.